### PR TITLE
Use static CUDA upstream libs on Unix

### PR DESCRIPTION
**Pull request by 9prady9**
- Use static cufft, cublas, cusolver and cusolver on Unix. Dynamic libraries are used on Windows.  (Check commit message for details)
- Removed cuda_thrust_sort_by_key static dependency (Check commit message for details)

### CODE REVIEW

**Comment by 9prady9** [comment]
Except the following commits
- `Renamed cu to cpp where possible and cleanup CUDA fast/orb`
- `move CUDA kernels to runtime(nvrtc) compilation where ever possible`

rest of them are actually fixes to enable linking with static cuda libs - it also fixes CUDA separable compilation in ArrayFire(broken right now) which is needed for the former to work successfully. 

The reason I kept them separate is those changes are independent enough to be a commit by themselves - change doesn't break non-separable compilation and tests. Each of the commits compile and pass tests on my local Linux and Windows box. Also, each commit is separated based on the type of change so that reviewing is easy. Moving them into different branches and there by other PRs doesn't add any value in this case I think.

Note: Except the nvrtc related commit, rest of them are largely CMake related changes barring the LUT commit (it is a minor change compared to nvrtc).

File: src/backend/cuda/culibs/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    cufft.hpp
+    fft.cu
+    sparse.cu
+    sparse.hpp
+    sparse_arith.cu
+    sparse_arith.hpp
+    sparse_blas.cu
+    sparse_blas.hpp
+    solve.cu
+    solve.hpp
```
Comment (umar456) [inline src/backend/cuda/culibs/CMakeLists.txt]: There is no context why these files are used in this cmake file. I don't like this approach. If you want to keep the cmake separate add this file to the CMakeModules folder in the root arrayfire directory create a function that adds these files with an explanation.

File: src/backend/cuda/diff.cpp [hunk 2]
```diff
@@ -15,8 +15,8 @@
 
 namespace cuda {
 
-template<typename T, bool isDiff2>
-static Array<T> diff(const Array<T> &in, const int dim) {
+template<typename T>
+Array<T> diff(const Array<T> &in, const int dim, const bool isDiff2) {
```
Comment (umar456) [inline src/backend/cuda/diff.cpp]: It would have been nice if these changes were done in a separate PR.

File: src/backend/cuda/kernel/memcopy.hpp [hunk 3]
```diff
@@ -237,12 +90,13 @@ void copy(Param<outType> dst, CParam<inType> src, int ndims,
         ((src.dims[0] == dst.dims[0]) && (src.dims[1] == dst.dims[1]) &&
          (src.dims[2] == dst.dims[2]) && (src.dims[3] == dst.dims[3]));
 
-    if (same_dims)
-        CUDA_LAUNCH((copy_kernel<inType, outType, true>), blocks, threads, dst,
-                    src, default_value, factor, trgt_dims, blk_x, blk_y);
-    else
-        CUDA_LAUNCH((copy_kernel<inType, outType, false>), blocks, threads, dst,
-                    src, default_value, factor, trgt_dims, blk_x, blk_y);
+    auto cpy = getKernel("cuda::copy", source,
```
Comment (umar456) [inline src/backend/cuda/kernel/memcopy.hpp]:
```suggestion
    auto copy = getKernel("cuda::copy", source,
```

File: src/backend/cuda/kernel/scan_by_key/CMakeLists.txt [hunk 4]
```diff
@@ -14,32 +14,15 @@ foreach(STR ${FILESTRINGS})
...
-    cuda_compile(scan_by_key_gen_files "${CMAKE_CURRENT_BINARY_DIR}/kernel/scan_by_key/scan_by_key_impl_${SBK_BINARY_OP}.cu"
-          "${CMAKE_CURRENT_SOURCE_DIR}/kernel/scan_dim_by_key_impl.hpp"
-          "${CMAKE_CURRENT_SOURCE_DIR}/kernel/scan_first_by_key_impl.hpp"
-          OPTIONS
-          -DSBK_BINARY_OP=${SBK_BINARY_OP} "${platform_flags} ${cuda_cxx_flags} -DAFDLL"
-      )
-
-    list(APPEND SCAN_OBJ ${scan_by_key_gen_files})
-endforeach(SBK_BINARY_OP ${SBK_BINARY_OPS})
+  configure_file(
```
Comment (umar456) [inline src/backend/cuda/kernel/scan_by_key/CMakeLists.txt]: Couldn't we just add cuda_scan_by_key target to the static library?

File: src/backend/cuda/culibs/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  mark_as_advanced(CUDA_${libname}_LIBRARY)
+  endmacro()
+
+  af_find_static_cuda_libs(culibos)
+  af_find_static_cuda_libs(cublas_static)
+  af_find_static_cuda_libs(cublasLt_static)
+  af_find_static_cuda_libs(cufft_static)
+  af_find_static_cuda_libs(cusparse_static)
+
+  set(extra_nvcc_flags "-rdc=true;-dlink;-lculibos")
```
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/cuda/culibs/CMakeLists.txt]: CUDA libraries path is handled by `cuda_*` cmake commands/macros. So, just doing this works fine.

File: src/backend/cuda/culibs/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  set(extra_nvcc_flags "${extra_nvcc_flags};-lcublas_static;-lcublasLt_static")
```
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/cuda/culibs/CMakeLists.txt]: CUDA libraries path is handled by `cuda_*` cmake commands/macros. So, just doing this works fine.

File: src/backend/cuda/culibs/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      ${CMAKE_DL_LIBS}
+      $<$<NOT:$<BOOL:${AF_USE_STATIC_CUDA_LAPACK}>>:${CUDA_cusolver_LIBRARY}>
+      -Wl,--start-group
+      ${CUDA_culibos_LIBRARY} #also a static libary
+      ${CUDA_cublas_static_LIBRARY}
+      ${CUDA_cublasLt_static_LIBRARY}
+      ${CUDA_cufft_static_LIBRARY}
+      ${CUDA_lapack_static_LIBRARY}
+      ${CUDA_cusparse_static_LIBRARY}
+      $<$<BOOL:${AF_USE_STATIC_CUDA_LAPACK}>:${CUDA_cusolver_static_LIBRARY}>
```
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/cuda/culibs/CMakeLists.txt]: I am not sure how that would work without duplication of these statements into larger if-else block. The static library has to appear inside `grouping` flags while the shared library outside to reflect what is happening there correctly.

File: src/backend/cuda/fast.cu [hunk 8]
```diff
@@ -7,17 +7,53 @@
...
-#include <Array.hpp>
-#include <err_cuda.hpp>
+#include <fast.hpp>
+
 #include <kernel/fast.hpp>
+#include <kernel/fast_lut.hpp>
 #include <af/dim4.hpp>
-#include <af/features.h>
+
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/cuda/fast.cu]: Will take care of it while picking commits into different PRs.

File: src/backend/cuda/CMakeLists.txt [hunk 9]
```diff
@@ -10,6 +10,18 @@ include(select_compute_arch)
 
 dependency_check(CUDA_FOUND "CUDA not found.")
 
+# Remove cublas_device library which is no longer included with the cuda
+# toolkit. Fixes issues with older CMake versions
+if(DEFINED CUDA_cublas_device_LIBRARY AND NOT CUDA_cublas_device_LIBRARY)
+  list(REMOVE_ITEM CUDA_CUBLAS_LIBRARIES ${CUDA_cublas_device_LIBRARY})
+endif()
+
+# Remove cublas_device library which is no longer included with the cuda
```
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/cuda/CMakeLists.txt]: I merely moved these to the top so that those changes are used by all cuda_add_library calls consistently. Yeah, they look the same, I didn't realize that while moving. Must have been the result of a three-way merge perhaps.

I will remove it in a later cleanup commit.

File: src/backend/common/internal_enums.hpp [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#pragma once
+
+typedef enum {
+    AF_BATCH_UNSUPPORTED = -1, /* invalid inputs */
```
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/common/internal_enums.hpp]: I will add it in corresponding PR while splitting.

File: src/backend/cuda/CMakeLists.txt [hunk 11]
```diff
@@ -473,60 +498,70 @@ cuda_add_library(afcuda
...
+
+    OPTIONS
+    ${platform_flags}
+    ${cuda_cxx_flags}
+    -Xcudafe \"--diag_suppress=1427\"
   )
 
 arrayfire_set_default_cxx_flags(afcuda)
 
-# NOTE: Do not add additional CUDA specific definitions here. Add it to the
```
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/cuda/CMakeLists.txt]: There are no other targets towards end of my final commit in this PR. I didn't see a point of retaining an obsolete comment.

File: src/backend/cuda/culibs/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    # contains GPU accelerated stedc and bdsqr. The user has to link
+    # libcusolver_static.a with liblapack_static.a in order to build the
+    # application successfully.
+    af_find_static_cuda_libs(lapack_static)
+
+    set(extra_nvcc_flags "${extra_nvcc_flags};-lcusolver_static")
+  endif()
+endif()
+
+cuda_add_library(cuda_culibs STATIC
```
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/cuda/culibs/CMakeLists.txt]: I didn't spend too much time wondering about this name, I guessed it would be changed.

File: src/backend/cuda/diff.cpp (hunk 2 above)
Comment (9prady9) [reply src/backend/cuda/diff.cpp]: These are part of nvrtc commit, a single commit. You are just seeing them clubbed into single change as you are using `Changes` button on PR directly instead of checking out changes by commit.

Since I am going to move this into separate PR, I am going to mark this resolved.

File: src/backend/common/internal_enums.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/common/internal_enums.hpp]: I will add it in corresponding PR while splitting.

File: src/backend/cuda/kernel/scan_by_key/CMakeLists.txt (hunk 4 above)
Comment (9prady9) [reply src/backend/cuda/kernel/scan_by_key/CMakeLists.txt]: I don't understand quite what you mean. The generated cpp files can be directly included into afcuda target and they don't take any time at all since it is going through nvcc anymore. It doesn't make sense to retain this old approach.

File: src/backend/cuda/CMakeLists.txt [hunk 13]
```diff
@@ -473,60 +498,70 @@ cuda_add_library(afcuda
...
+    wrap.cpp
     wrap.hpp
 
     jit/BufferNode.hpp
     jit/kernel_generators.hpp
 
     nvrtc/cache.cpp
 
-    OPTIONS ${platform_flags} ${cuda_cxx_flags} -Xcudafe \"--diag_suppress=1427\"
+    ${scan_by_key_sources}
```
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/cuda/CMakeLists.txt]: Yes, what you said is good for thrust sources. scan by key is fast now since they don't go through nvcc.

File: src/backend/cuda/culibs/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+  # lapack library. Otherwise, we have to use regular shared library
+  if(CUDA_VERSION_MAJOR VERSION_GREATER 10 OR CUDA_VERSION_MAJOR VERSION_EQUAL 10)
+    set(AF_USE_STATIC_CUDA_LAPACK ON CACHE INTERNAL "")
+  else()
+    set(AF_USE_STATIC_CUDA_LAPACK OFF CACHE INTERNAL "")
+  endif()
+
+  # The following macro uses a macro defined by
+  # FindCUDA module from cmake.
+  macro(af_find_static_cuda_libs libname)
```
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/cuda/culibs/CMakeLists.txt]: what is the downside to the current way ? I didn't want to define one for OS where it is not used at all. Hence, it is wrapped in the if block with everything else.

File: src/backend/cuda/culibs/CMakeLists.txt (hunk 1 above)
Comment (9prady9) [reply src/backend/cuda/culibs/CMakeLists.txt]: I am not saying what I have now is optimum approach, but putting it in CMakeModules isn't correct. This section of code is very specific to CUDA backend and much like how we had `cuda_thrust_sort_by_key` and `cuda_scan_by_key` targets, this also is including sources from `src/backend/cuda`. Therefore, I think this file should be under `src/backend/cuda`. I have no specific opinion as to where under `src/backend/cuda`, anywhere would work as long as we `include` it properly.

File: src/backend/cuda/dims_param.hpp [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#pragma once
+
+namespace cuda {
+
+typedef struct {
+    int dim[4];
+} dims_t;
```
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/cuda/dims_param.hpp]: Yes, that was correct until recently. I am using Param structure now with memcpy and copy kernels and the tests pass fine without any issues. However, this `dims_t` is used else where in `kernel::copy` function to pass some extra dimensions (apart from Param/CParam) - an array of 4 dim_t elements.

I didn't have a strong opinion of retaining vs removing it, since there could be some future cases where such `dims_t` POD struct can be used by other kernels too. So, I kept it.

File: src/backend/common/internal_enums.hpp [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    AF_BATCH_NONE,             /* one signal, one filter   */
+    AF_BATCH_LHS,              /* many signal, one filter  */
+    AF_BATCH_RHS,              /* one signal, many filter  */
+    AF_BATCH_SAME,             /* signal and filter have same batch size */
+    AF_BATCH_DIFF,             /* signal and filter have different batch size */
+} AF_BATCH_KIND;
+
+enum class kJITHeuristics {
```
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/common/internal_enums.hpp]: Taking care of it in the new PR. Going to raise one soon.

File: src/backend/cuda/culibs/CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (umar456) [reply src/backend/cuda/culibs/CMakeLists.txt]:
```
if(AF_USE_STATIC_CUDA_LAPACK)
     target_link_libraries(cuda_culibs
    PRIVATE
        ${CUDA_cusolver_static_LIBRARY}
else()
     target_link_libraries(cuda_culibs
    PRIVATE
        ${CUDA_cusolver_LIBRARY})
endif()
```

File: src/backend/cuda/CMakeLists.txt (hunk 11 above)
[the remark this answers is not in the document]
Comment (umar456) [reply src/backend/cuda/CMakeLists.txt]: The idea was to make sure people don't append additional flags to the end of this statement. For example `target_compile_options(afcuda PRIVATE ${cxx_definitions};-my-super-awesome-flag)`

File: src/backend/cuda/dims_param.hpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (umar456) [reply src/backend/cuda/dims_param.hpp]: I feel like this encourages its use which we should avoid. I think we should remove it from its own header.

File: src/backend/cuda/kernel/fast_lut.hpp [hunk 17]
```diff
@@ -1,13 +1,15 @@
...
  *
  * This file is distributed under 3-clause BSD license.
  * The complete license agreement can be obtained at:
  * http://arrayfire.com/licenses/BSD-3-Clause
  ********************************************************/
 
-__constant__ unsigned char FAST_LUT[] = {
+#pragma once
+
+unsigned char FAST_LUT[] = {
```
Comment (umar456) [inline src/backend/cuda/kernel/fast_lut.hpp]: This array makes me sad.

File: src/backend/cuda/CMakeLists.txt (hunk 11 above)
[the remark this answers is not in the document]
Comment (9prady9) [reply src/backend/cuda/CMakeLists.txt]: adding it back in rebase.

**Comment by daniel-dsouza** [comment]
Is there still a way to dynamically link to CUDA? The uncompressed libafcuda.so.3.7.2 is almost a gigabyte in size!

**Comment by 9prady9** [comment]
@daniel-dsouza The 1GB size of binary is not due to static linking of dynamic libs. It is rather due to compiling the library with multiple compute versions starting from 30 to 75 of CUDA cards. For our website binaries to work optimally on all CUDA cards, we build with all compute versions, hence increasing the binary size. For example, for a single compute version, the CUDA backend will end up being approx 450MB. Having said that, we are aware of the issue with such huge size and we are constantly working on reducing it. As part of that effort, we moved most of the kernels for runtime compilation instead of offline compilation using nvcc. This brought down CUDA backend binary a good bit. We have more work in the pipeline towards the same goal for future releases.

**Comment by daniel-dsouza** [comment]
@9prady9 Thank you for explaining this to me. Currently I am building from compute version 5 and higher, so I can look into narrowing that down to my specific hardware.
